File: javaep/TaskerQueue.java
package multask;

import multask.Tasker;

public class TaskerQueue {
	private class TaskerItem{
		public Tasker tasker=null;
		public TaskerItem last_tasker_item=null;
		public TaskerItem next_tasker_item=null;
		
		public TaskerItem(Tasker tasker){
			this.tasker=tasker;
		}
	}
	
	private TaskerItem tasker_queue_head=null;
	private TaskerItem tasker_queue_tail=null;
	private Object read_write_lock=null;
	private Object read_lock=null;
	private Object write_lock=null;
	private int count=0;
	
	public TaskerQueue(){
		this.tasker_queue_head=new TaskerItem(null);
		this.tasker_queue_tail=new TaskerItem(null);
		this.tasker_queue_head.next_tasker_item=this.tasker_queue_tail;
		this.tasker_queue_tail.last_tasker_item=this.tasker_queue_head;
		this.read_write_lock=new Object();
		this.read_lock=new Object();
		this.write_lock=new Object();
	}
	
	public void push(Tasker tasker){
		TaskerItem ti=new TaskerItem(tasker);
		this.tasker_queue_tail.last_tasker_item.next_tasker_item=ti;
		ti.last_tasker_item=this.tasker_queue_tail.last_tasker_item;
		this.tasker_queue_tail.last_tasker_item=ti;
		ti.next_tasker_item=this.tasker_queue_tail;
		this.count++;
	}
	
	public Tasker pop(){
		if(this.count<=0) return null;
		TaskerItem ti=this.tasker_queue_head.next_tasker_item;
		ti.next_tasker_item.last_tasker_item=this.tasker_queue_head;
		this.tasker_queue_head.next_tasker_item=ti.next_tasker_item;
		ti.last_tasker_item=null;
		ti.next_tasker_item=null;
		this.count--;
		return ti.tasker;
	}
	
	public int count(){
		return this.count;
	}
}

File: cep/TaskerQueue.cpp
#include <Windows.h>
#include "multask.h"
#include "tasker.h"
#include "taskerqueue.h"

CRITICAL_SECTION link_cs;
Tasker* head;
Tasker* tail;
long count;

void enterCriticalSection(){
	long time_s=GetCurrentTime();
	while (!TryEnterCriticalSection(&link_cs))
	{
		if(GetCurrentTime()-time_s>1){
			EnterCriticalSection(&link_cs);
			break;
		}
	}
}

void multask_initTaskerQueue(){
	multask_initTasker();
	InitializeCriticalSection(&link_cs);
	head=multask_createTasker("__tasker_queue_head__",0,NULL,NULL);
	tail=multask_createTasker("__tasker_queue_tail__",0,NULL,NULL);
	head->next=tail;
	tail->last=head;
	count=0;
}

void multask_unInitTaskerQueue(){
	Tasker* ti=head;
	Tasker* tit=NULL;
	while(ti->next!=NULL){
		tit=ti;
		ti=ti->next;
		multask_freeTasker(tit);
	}
	multask_unInitTasker();
	DeleteCriticalSection(&link_cs);
}

void multask_taskerPush(Tasker* tasker){
	enterCriticalSection();
	tasker->last=tail->last;
	tasker->next=tail;
	tail->last->next=tasker;
	tail->last=tasker;
	count++;
	LeaveCriticalSection(&link_cs);
}

Tasker* multask_taskerPop(){
	Tasker* t=NULL;
	enterCriticalSection();
	if(count>0){
		t=head->next;
		t->next->last=head;
		head->next=t->next;
		count--;
	}
	LeaveCriticalSection(&link_cs);
	if(t!=NULL){
		t->last=NULL;
		t->last=NULL;
	}
	return t;
}

long multask_taskerQueueCount(){
	return count;
}